TDL15-CloudFormation-Jenkins/cf-deploy-jenkins.yaml

```diff
@@ -0,0 +1,245 @@
+          # set permissions
+          sudo chown -R jenkins:jenkins /home/jenkins /var/lib/jenkins /var/cache/jenkins /var/log/jenkins; sudo chmod 0700 /var/lib/jenkins/.ssh; sudo chmod 0600 /var/lib/jenkins/.ssh/*
+          # firewall: 
+          sudo firewall-cmd --permanent --add-service=ssh; sudo firewall-cmd --permanent --add-port='${JenkinsPort}'/tcp; sudo firewall-cmd --reload
+          # start the Jenkins service and enable it to start on system boot
+          sudo systemctl start jenkins
+          sudo systemctl enable jenkins
+          # install git
+          yum install -y git
+          # install AWS CLI
+          curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip"
+          unzip awscliv2.zip
+          sudo ./aws/install
+          # install python
+          sudo yum -y groupinstall "Development Tools"
+          # sudo yum -y install openssl-devel bzip2-devel libffi-devel
+          sudo yum -y install bzip2-devel expat-devel gdbm-devel \
+            ncurses-devel openssl-devel readline-devel wget \
+            sqlite-devel tk-devel xz-devel zlib-devel libffi-devel
+          sudo wget https://www.python.org/ftp/python/${PythonVersion}/Python-${PythonVersion}.tgz
+          tar -xf Python-${PythonVersion}.tgz
+          cd Python-${PythonVersion}
+          ./configure --enable-optimizations # Setup installation by running the configure script, performs a number of checks to make sure all of the dependencies on your system are present
+          make -j 4 # Start the Python 3.8 build process 
+          sudo make altinstall #install the Python binaries
+          alias python="/usr/local/bin/python3.8"
+          # intall boto3
+          pip3.8 install boto3
+          # install terraform 
+          sudo wget https://releases.hashicorp.com/terraform/${TerraformVersion}/terraform_${TerraformVersion}_linux_amd64.zip
+          unzip terraform_${TerraformVersion}_linux_amd64.zip && sudo rm terraform_${TerraformVersion}_linux_amd64.zip -f && sudo mv terraform /usr/local/bin/
+
+  # Define the Security Group that will be appended to the ENI of the Instance.  
+  ServerSecurityGroup:
+    Type: AWS::EC2::SecurityGroup
+    Properties:
+      GroupDescription: Security Group that will be used for the Jenkins instance. Open ports 22, and the JenkinsPort
+      VpcId: !Ref VpcId
+      SecurityGroupIngress:
+        - IpProtocol: tcp
+          FromPort: 22
+          ToPort: 22
+          CidrIp: !Ref SSHLocation
+        - IpProtocol: tcp
+          FromPort: !Ref JenkinsPort
+          ToPort: !Ref JenkinsPort
+          CidrIp: !Ref SSHLocation       
+
+  # Create the EBS volume that we will use for /var/lib/jenkins
+  JenkinsDataVolume:
+    Type: AWS::EC2::Volume
+    Properties:
+      Size: !Ref VolumeSize
+      SnapshotId:  !If [UseSnapshot, !Ref SnapshotId, !Ref "AWS::NoValue"]
+      Encrypted: true
```

Comment:
For encrypted disk, it is better to provide customized KMS instead of the default KMS key.

Reply:
Good advice,thanks.